### PR TITLE
Code Review: startup.m2, Core.m2, and last.m2

### PR DESCRIPTION
**Pull request by mahrud**
This is _not_ a pull request, but rather a hack to use GitHub's features for performing a collaborative code review. If this works well, I'm hoping to also look at parts of the interpreter later. Alternatively, if anyone knows of a collaborative tool for reviewing existing code (i.e. not reviewing a diff but a whole file), that would be very welcome!

Specifically with an eye towards future pull requests and refactoring, my goals are to:
1. go through each section/variable/function in the three files,
2. explain what is the purpose of that code is/was, and
3. discuss whether
    - that purpose is still necessary, otherwise delete the section, or
    - we would do it differently or more effectively now.

Basically imagine if someone submitted a pull request to add these files and your task is to be the reviewer.

Here is how to participate:
1. click on "Files changed" tab above
2. click on "Load diff" for any of the files
3. read any part of the code and if you have questions about it, leave a comment
4. answer any questions that you are familiar with below

Tagging people who have changed these files: @DanGrayson @mikestillman @d-torrance @pzinn

### CODE REVIEW

File: M2/Macaulay2/m2/startup.m2.in [hunk 1]
```diff
@@ -1,642 +0,0 @@
...
-notify = false
-debugWarningHashcode = null
-gotarg := arg -> any(commandLine, s -> s === arg)
-randomSeed = -- initializing this here so the startup is reproducible
-if gotarg "--no-randomize" then 0 else (currentTime() << 16) + processID()
-if gotarg "--notify" then notify = true
-if gotarg "--stop" then stopIfError = true
-if gotarg "--int"  then handleInterrupts = false
-
-firstTime := class PackageDictionary === Symbol
```
Comment (mahrud) [inline M2/Macaulay2/m2/startup.m2.in:19]: I vaguely recall that at some point a running M2 process could be halted and "pickled" into a core dump in such a way that it could be restarted from that position, and the point of `firstTime` was to indicate whether we're loading for the first time or from a pickle. Is this correct?

Should we remove it everywhere?

File: M2/Macaulay2/m2/startup.m2.in [hunk 2]
```diff
@@ -1,642 +0,0 @@
...
-    printerr = msg -> endl(printString(stderr, concatenate(loadDepth, "-- ", msg)));
-    value = value'; -- TODO: is there a different way to get --no-core to work?
-    --
-    PackageDictionary = new Dictionary;
-    dictionaryPath = append(dictionaryPath, PackageDictionary);
-    assert'( not isGlobalSymbol "Core" );
-    PackageDictionary#("Package$Core") = getGlobalSymbol(PackageDictionary, "Core");
-    )
--- we can't make this an else-clause, because then "Core" will be in the wrong dictionary
-disassemble ( () -> debug Core )
```
Comment (mahrud) [inline M2/Macaulay2/m2/startup.m2.in:37]: What does this line do?

File: M2/Macaulay2/m2/startup.m2.in [hunk 3]
```diff
@@ -1,642 +0,0 @@
...
-if not firstTime then debug Core -- we need access to the private symbols (we remove the Core private dictionary later.)
-
-toString := value' getGlobalSymbol if firstTime then "simpleToString" else "toString"
-
-local exe
-local topBuilddir
```
Comment (mahrud) [inline M2/Macaulay2/m2/startup.m2.in:43]: What's the point of declaring these so early? Why not just say `exe = commandLine#0`?

File: M2/Macaulay2/m2/startup.m2.in (hunk 2 above)
Comment (mahrud) [inline M2/Macaulay2/m2/startup.m2.in:37]: What does this line do?

File: M2/Macaulay2/m2/startup.m2.in [hunk 4]
```diff
@@ -1,642 +0,0 @@
...
-	 (true, ret)) else (false, null);
-     trySimpleLoad := (filename, filepath) -> first tryLoad(filename, filepath, simpleLoad, notify);
-
-     normalPrompts = () -> (
-	  lastprompt := "";
-	  ZZ#{Standard,InputPrompt} = lineno -> concatenate(newline, lastprompt = concatenate(interpreterDepth:"i", toString lineno, " : "));
-	  ZZ#{Standard,InputContinuationPrompt} = lineno -> #lastprompt; -- will print that many blanks, see interp.d
-	  symbol currentPrompts <- normalPrompts;	    -- this avoids the warning about redefining a function
-	  );
-     normalPrompts();
```
Comment (mahrud) [inline M2/Macaulay2/m2/startup.m2.in:82]: This function runs here, then again in line 592. Why?

File: M2/Macaulay2/m2/startup.m2.in [hunk 5]
```diff
@@ -1,642 +0,0 @@
...
-nocore = false
-noinitfile = false
-interpreter := commandInterpreter
-
--*
-getRealPath := fn -> (					    -- use this later if realpath doesn't work
-     local s;
-     while ( s = readlink fn; s =!= null ) do fn = if isAbsolutePath s then s else minimizeFilename(fn|"/../"|s);
-     fn)
-*-
```
Comment (mahrud) [inline M2/Macaulay2/m2/startup.m2.in:342]: Is this dead code?

File: M2/Macaulay2/m2/startup.m2.in [hunk 6]
```diff
@@ -1,642 +0,0 @@
...
-
-showMaps := () -> (
-     if version#"operating system" === "SunOS" then (
-	  stack lines get ("!/usr/bin/pmap "|processID())
-	  )
-     else if version#"operating system" === "Linux" and fileExists("/proc/"|toString processID()|"/maps") then (
-	  stack lines get("/proc/"|toString processID()|"/maps")
-	  )
-     else "memory maps not available"
-     )
```
Comment (mahrud) [inline M2/Macaulay2/m2/startup.m2.in:413]: What's the point of `showMaps`? Something to do with pickling?

**Comment by mahrud** [comment]
@DanGrayson do you have time to help answer any of the questions here?

**Comment by DanGrayson** [comment]
> @DanGrayson do you have time to help answer any of the questions here?

No, sorry!

**Comment by mahrud** [comment]
Seems like code review is not a priority.
